pkg/protocol/block_dispatcher.go

```diff
@@ -0,0 +1,345 @@
+	if err != nil {
+		return ierrors.Wrapf(err, "failed to get slot %d (not committed yet)", commitmentID.Index())
+	}
+
+	commitment, err := committedSlot.Commitment()
+	if err != nil {
+		return ierrors.Wrapf(err, "failed to get commitment from slot %d", commitmentID.Index())
+	} else if commitment.ID() != commitmentID {
+		return ierrors.Wrapf(err, "commitment ID mismatch: %s != %s", commitment.ID(), commitmentID)
+	}
+
+	blockIDs, err := committedSlot.BlockIDs()
+	if err != nil {
+		return ierrors.Wrapf(err, "failed to get block IDs from slot %d", commitmentID.Index())
+	}
+
+	roots, err := committedSlot.Roots()
+	if err != nil {
+		return ierrors.Wrapf(err, "failed to get roots from slot %d", commitmentID.Index())
+	}
+
+	b.protocol.networkProtocol.SendWarpSyncResponse(commitmentID, blockIDs, roots.TangleProof(), src)
+
+	return nil
+}
+
+// processWarpSyncResponse processes a WarpSync response.
+func (b *BlockDispatcher) processWarpSyncResponse(commitmentID iotago.CommitmentID, blockIDs iotago.BlockIDs, merkleProof *merklehasher.Proof[iotago.Identifier], _ network.PeerID) error {
+	if b.processedWarpSyncRequests.Has(commitmentID) {
+		return nil
+	}
+
+	chainCommitment, exists := b.protocol.ChainManager.Commitment(commitmentID)
+	if !exists {
+		return ierrors.Errorf("failed to get chain commitment for %s", commitmentID)
+	}
+
+	targetEngine := b.targetEngine(chainCommitment)
+	if targetEngine == nil {
+		return ierrors.Errorf("failed to get target engine for %s", commitmentID)
+	}
+
+	acceptedBlocks := ads.NewSet[iotago.BlockID](mapdb.NewMapDB(), iotago.BlockID.Bytes, iotago.SlotIdentifierFromBytes)
+	for _, blockID := range blockIDs {
+		_ = acceptedBlocks.Add(blockID) // a mapdb can newer return an error
+	}
+
+	if !iotago.VerifyProof(merkleProof, iotago.Identifier(acceptedBlocks.Root()), chainCommitment.Commitment().RootsID()) {
+		return ierrors.Errorf("failed to verify merkle proof for %s", commitmentID)
+	}
+
+	b.pendingWarpSyncRequests.StopTicker(commitmentID)
+
+	b.processedWarpSyncRequests.Add(commitmentID)
+
+	targetEngine.Events.BlockDAG.BlockSolid.Hook(func(block *blocks.Block) {
+		block.ID()
+	})
+
+	for _, blockID := range blockIDs {
+		targetEngine.BlockDAG.GetOrRequestBlock(blockID)
+	}
+
+	return nil
+}
+
+// inWarpSyncRange returns whether the given block should be processed by a warp sync process.
+//
+// This is the case if the block is more than a warp sync threshold ahead of the latest commitment while also committing
+// to a new slot that can be warp synced.
+func (b *BlockDispatcher) inWarpSyncRange(engine *engine.Engine, block *model.Block) bool {
+	if engine.BlockRequester.HasTicker(block.ID()) {
+		return false
+	}
+
+	slotCommitmentID := block.ProtocolBlock().SlotCommitmentID
+	latestCommitmentIndex := engine.Storage.Settings().LatestCommitment().Index()
+	maxCommittableAge := engine.APIForSlot(slotCommitmentID.Index()).ProtocolParameters().MaxCommittableAge()
```

Comment:
Why don't we use the max of `warpSyncWindowSize := lo.Cond(maxCommittableAge > b.optWarpSyncWindowSize, maxCommittableAge, b.optWarpSyncWindowSize)` these two here?